src/SignalR/clients/ts/signalr/src/HubConnection.ts

```diff
@@ -101,48 +118,110 @@ export class HubConnection {
 
+        try {
+            await this.startInternal();
+
+            this.connectionState = HubConnectionState.Connected;
+            this.logger.log(LogLevel.Debug, "HubConnection connected successfully.");
+        } catch (e) {
+            this.connectionState = HubConnectionState.Disconnected;
+            this.logger.log(LogLevel.Debug, `HubConnection failed to start successfully because of error '${e}'.`);
+            return Promise.reject(e);
+        }
+    }
+
+    public async startInternal() {
         this.receivedHandshakeResponse = false;
-        // Set up the promise before any connection is started otherwise it could race with received messages
+        // Set up the promise before any connection is (re)started otherwise it could race with received messages
         const handshakePromise = new Promise((resolve, reject) => {
             this.handshakeResolver = resolve;
             this.handshakeRejecter = reject;
         });
 
         await this.connection.start(this.protocol.transferFormat);
 
-        this.logger.log(LogLevel.Debug, "Sending handshake request.");
+        try {
+            const handshakeRequest: HandshakeRequestMessage = {
+                protocol: this.protocol.name,
+                version: this.protocol.version,
+            };
 
-        await this.sendMessage(this.handshakeProtocol.writeHandshakeRequest(handshakeRequest));
+            this.logger.log(LogLevel.Debug, "Sending handshake request.");
 
-        this.logger.log(LogLevel.Information, `Using HubProtocol '${this.protocol.name}'.`);
+            await this.sendMessage(this.handshakeProtocol.writeHandshakeRequest(handshakeRequest));
 
-        // defensively cleanup timeout in case we receive a message from the server before we finish start
-        this.cleanupTimeout();
-        this.resetTimeoutPeriod();
-        this.resetKeepAliveInterval();
+            this.logger.log(LogLevel.Information, `Using HubProtocol '${this.protocol.name}'.`);
+
+            // defensively cleanup timeout in case we receive a message from the server before we finish start
+            this.cleanupTimeout();
+            this.resetTimeoutPeriod();
+            this.resetKeepAliveInterval();
 
-        // Wait for the handshake to complete before marking connection as connected
-        await handshakePromise;
-        this.connectionState = HubConnectionState.Connected;
+            await handshakePromise;
+        } catch (e) {
+            this.logger.log(LogLevel.Debug, `Hub handshake failed with error '${e}' during start(). Stopping HubConnection.`);
+
+            this.cleanupTimeout();
+            this.cleanupPingTimer();
+
+            // HttpConnection.stop() should not complete until after the onclose callback is invoked.
+            // This will transition the HubConnection to the disconnected state before HttpConnection.stop() completes.
+            await this.connection.stop(e);
+            return Promise.reject(e);
```

Comment:
Because this is in an async method, should this be a throw instead of `Promise.reject(e);`